### ISSUE TITLE
AutoTriage: GitHub Issue and PR Assistant

### ISSUE THREAD

**Issue by nikhilsna** [shriya1401/Optimizers_pages#5]
<html><head></head><body><h1>AutoTriage: GitHub Issue &amp; PR Assistant</h1>
<h2>Problem</h2>
<p>Our OCS repo has 47 open issues. Nobody knows what's a bug vs feature, what's urgent, which files are affected, or who should work on it. We waste 10+ minutes per issue doing manual triage. Then Mortensen reviews our PRs and spends time pointing out missing JavaDocs instead of discussing design.</p>
<p><strong>This bot automates the boring parts so we focus on building.</strong></p>
<hr>
---
## Problem Statement
(Add problem statement with clear plan)
---
<h2>What It Does</h2>
<h3>1. Auto-Triages Issues</h3>
<p>Someone creates: "Login button doesn't work on mobile"</p>
<p>Bot comments within 10 seconds:</p>
<pre><code>🏷️ Classified as: bug, frontend, priority:high (confidence: 89%)

📂 Likely affected files:
  • LoginButton.jsx (similarity: 92%)
  • auth.js (similarity: 78%)

👤 Suggested assignee: @shriya (last modified LoginButton.jsx)

📊 Estimated effort: 2-3 hours (similar to issue #31)
</code></pre>
<h3>2. Reviews PRs Before Mortensen Sees Them</h3>
<p>Someone opens PR with messy code.</p>
<p>Bot comments immediately:</p>
<pre><code>⚠️ Code Review Findings:

Style (3 issues):
  • Missing JavaDoc on getUserData()
  • Inconsistent indentation

Code Smells (2 issues):
  • searchHandler() is 87 lines (recommended: &lt;50)
  • Nested if-statements 4 levels deep

Security (1 issue):
  • Potential SQL injection on line 45

Test Coverage: 45% (target: 70%)
</code></pre>
<p>We fix it ourselves. Mortensen sees clean code.</p>
<hr>
<h2>How It Works</h2>
<pre><code>GitHub Issue/PR Created
    ↓
Webhook hits our FastAPI server (Railway)
    ↓
Bot runs:
  • Issue Classifier (Naive Bayes or Claude API)
  • File Matcher (TF-IDF similarity + git history)
  • Code Analyzer (ESLint + complexity metrics)
    ↓
Bot posts GitHub comment with findings
</code></pre>
<p><strong>We're integrating existing tools, not inventing them:</strong></p>
<ul>
<li>Classification: scikit-learn or Claude API</li>
<li>File matching: TF-IDF (50 lines of Python)</li>
<li>Code review: ESLint, Checkstyle, JaCoCo (already exist)</li>
<li>GitHub: Webhooks + REST API (well-documented)</li>
</ul>
<hr>
<h2>8-Week Build Plan</h2>

Week | Milestone | Deliverable
-- | -- | --
1-2 | Webhook integration | Create issue → bot comments "received"
3-4 | Issue classification | Create issue → bot auto-labels correctly (80%+ accuracy)
5-6 | File relevance | Create issue → bot suggests which files to check
7 | Code review | Open PR → bot flags style/complexity/coverage issues
8 | Polish + testing | Deploy to real OCS repo, measure accuracy

<p><strong>Checkpoint (Week 4):</strong> If classification accuracy &lt;80%, switch from scikit-learn to Claude API.</p>
<hr>
<h2>Demo (5 Minutes)</h2>
<p><strong>Minute 1:</strong> Show the problem - 47 open issues, no organization</p>
<p><strong>Minute 2:</strong> Live create issue during presentation</p>
<ul>
<li>Bot responds in 10 seconds with classification + file suggestions</li>
</ul>
<p><strong>Minute 3:</strong> Live open PR with bad code</p>
<ul>
<li>Bot immediately flags violations we should've caught</li>
</ul>
<p><strong>Minute 4:</strong> Show metrics from Week 1 usage</p>
<ul>
<li>12 issues triaged, 91% accuracy, 18 style violations caught</li>
</ul>
<p><strong>Minute 5:</strong> Q&amp;A</p>
<hr>
<h2>Why This Isn't Vaporware</h2>
<p><strong>"How does classification work?"</strong>
We train on our existing 200+ labeled issues. Naive Bayes classifier. If accuracy sucks, fall back to Claude API with few-shot prompting.</p>
<p><strong>"How do you know which files are relevant?"</strong>
TF-IDF vector similarity between issue text and file contents. Same technique Stack Overflow uses. Plus git history for recent contributors.</p>
<p><strong>"Code review is subjective. How does a bot do it?"</strong>
We're not doing subjective review. We're running ESLint (style), cyclomatic complexity (method length), and JaCoCo (test coverage). Automating the mechanical parts, not design critique.</p>
<p><strong>"What if it's wrong?"</strong>
It's a suggestion, not a command. Humans make final decisions. Low stakes.</p>
<p><strong>"Can you actually build this in 8 weeks?"</strong>
Yes. All the pieces exist. We're assembling, not inventing.</p>
<hr>
<h2>What We're NOT Building</h2>
<p>❌ Auto-closing issues (human decision)<br>
❌ Auto-merging PRs (dangerous)<br>
❌ Custom ML models from scratch (use scikit-learn)<br>
❌ Multi-repo support (OCS repo only for MVP)<br>
❌ Slack notifications (GitHub comments sufficient)</p>
<hr>
<h2>Success Metrics</h2>
<p><strong>Primary (Week 9 - first week of real use):</strong></p>
<ul>
<li>Classification accuracy ≥80%</li>
<li>File suggestions correct ≥70% of time</li>
<li>Bot catches ≥5 style violations per week</li>
<li>Time-to-triage: 12min → &lt;2min per issue</li>
</ul>
<p><strong>Deployment:</strong></p>
<ul>
<li>Deploy to actual OCS repo</li>
<li>Team uses it daily</li>
<li>If other teams want it, install as GitHub App</li>
</ul>
<hr>
<h2>Tech Stack</h2>
<ul>
<li><strong>Backend:</strong> FastAPI (Python) on Railway</li>
<li><strong>Storage:</strong> PostgreSQL (issue history) + Redis (caching)</li>
<li><strong>ML:</strong> scikit-learn (classification) or Claude API (fallback)</li>
<li><strong>Integration:</strong> GitHub Webhooks + REST API</li>
<li><strong>Code Analysis:</strong> ESLint, Checkstyle, JaCoCo</li>
</ul>
<p><strong>Cost:</strong> $5/month Railway Hobby tier (includes PostgreSQL + Redis)</p>
<hr>
<h2>Why This is the Right Project</h2>

Tracking system for soft skills

How do we know students are doing soft skills?

Have the bot triage - not human triage

We need implementation, because students do code all the time

People can make a blank issue and fool the system

Mort doesn’t want the noise of being in their social group - he needs the monitoring

Look at risha api issue
Issues as a backend to make sure people have key soft skills
Ppl who don’t do design and we start a sprint, they’re the ones who get stuck and they just try to wing it
They get 3-4 days in and they’re stuck
Without a Retrospective u have nothing to work on afterwards
The coding isn’t the problem, it’s just going forward that’s the issue
